### PR TITLE
Improve point conversions in layer hierarchy

### PR DESCRIPTION
**Pull request by jonastreub**
New matrices on layer to go from layer points to screen and canvas coordinates. Useful for bounding boxes etc.

`canvasFrame` and `screenFrame` now return the correct frames.

#### new properties on `Layer`
- `matrix`, WebKitCSSMatrix
- `transformMatrix`, WebKitCSSMatrix
- `matrix3d`, WebKitCSSMatrix
- `convertPointToScreen(point)` -> point
- `convertPointToCanvas(point)` -> point
- `convertPointFromScreen(point)` -> point
- `convertPointFromCanvas(point)` -> point

### CODE REVIEW

**Comment by 1note** [comment]
This is great! Does that mean it would help me make something like this?

http://share.framerjs.com/sn9q2hq3gbmi/

**Comment by jordandobson** [comment]
This is great! Does that mean it would help me make something like this?

http://share.framerjs.com/sn9q2hq3gbmi/

**Comment by jonastreub** [comment]
Sure! That should already work using the `screenFrame`. I am still fixing some issues translating to the canvas when perspective is at play.

File: framer/Layer.coffee [hunk 1]
```diff
@@ -209,6 +210,111 @@ class exports.Layer extends BaseClass
...
+			point.z = 0 if layer.flat
+			point = layer.matrix3d.point(point)
+			if !layer.superLayer
+				point.z = 0
+
+		point
+
+	_fromScreenPoint: (point = {}, context = false) =>
+		point =
+			x: point.x || 0
```
Comment (koenbok) [inline framer/Layer.coffee]: Lets use `or` instead of `||`

File: framer/Layer.coffee [hunk 2]
```diff
@@ -209,6 +210,111 @@ class exports.Layer extends BaseClass
...
+		for layer in @superLayers(context)
+			point.z = 0 if layer.flat
+			point = layer.matrix3d.point(point)
+			if !layer.superLayer
+				point.z = 0
+
+		point
+
+	_fromScreenPoint: (point = {}, context = false) =>
+		point =
```
Comment (koenbok) [inline framer/Layer.coffee]: You can use `_.defaults` for this: `point = `_.defaults(point, {x:0, y:0, z:0})`

File: framer/Layer.coffee [hunk 3]
```diff
@@ -209,6 +210,111 @@ class exports.Layer extends BaseClass
...
+				.translate(-@originX * @width, -@originY * @height)
+
+	_perspectiveProjectionMatrix: (element) =>
+		p = element.perspective
+		m = new Matrix()
+		m.m34 = -1/p if p? and p isnt 0
+		m
+
+	_perspectiveMatrix: (element) =>
+		if !element
```
Comment (koenbok) [inline framer/Layer.coffee]: `element ?= Screen`

**Comment by koenbok** [comment]
WIP:

- Some stylistic issues.
- We need documenting comments.
- We need to think about replacing the `Utils.convertPoint()` implementation.
- I think we need a better name for _boundingBox and maybe split it into a Utils function that takes the four corners of a frame and applies a function.

**Comment by koenbok** [comment]
This looks good @jonastreub. Please add a bit more descriptive comments in the Utils functions where still missing and add some unit tests for the Utils function. For complex cases, just do something in FS, verify it's correct visually, and copy the expected numbers to the test.

**Comment by jordandobson** [comment]
@jonastreub I'm not sure screenFrame is working how you'd expect.

Click on the line to toggle between using ScreenFrame and the function I implented to calculate the bounding box.

http://jrdn.io/1D1G0d0q0w3t

**Comment by jonastreub** [comment]
@jordandobson This pull request is not yet merged with the master branch. Here is an example showing both screenFrame and canvasFrame in sync: http://share.framerjs.com/jiw6k1nwrz0s/

**Comment by jordandobson** [comment]
Okay... sorry. I thought you were saying that it would already work! Looking forward to this! I will definitely use it. :smile:

**Comment by jonastreub** [comment]
It is going to be merged really soon!

**Pull request merged**